### PR TITLE
Get Haiku platform crate compiling on non-Haiku platforms

### PR DESCRIPTION
**Pull request by lopopolo**
Some changes to get `iana-time-zone-haiku`'s fallback code compiling and working correctly on non-Haiku hosts:

- Only link to `be` lib when the target is a Haiku target triple.
- Parse the target triple in `build.rs` from the `TARGET` env var which is set by Cargo.
- Fix cargo pragmas in `build.rs` to properly emit `rerun-if-changed` directives.
- Use C++11 namespace declaration syntax (the current `implementation.cc` file used C++17 syntax which I imagine was not caught because `-std=c++11` is only passed if supported by the host cxx compiler).
- Add CI to build and test `iana-time-zone-haiku` on macOS, Linux, and Windows.
- Add CI to run `cargo check` and `cargo clippy` on `iana-time-zone-haiku`.
- Add standard set of lints to `iana-time-zone-haiku`. Extracted from #76.
- Add `Cargo.toml` keywords and categories to `iana-time-zone-haiku`.
- Fix typo where crate level docs were incorrectly declared with `///` (changed to `//!`)
- Add docs to `iana_time_zone_haiku::get_timezone()`.
- Add example to `iana_time_zone_haiku::get_timezone()`.
- Add smoke test for `iana_time_zone_haiku::get_timezone()` to ensure the fallback routine returns `None` on all non-Haiku platforms.
- Add smoke test for `iana_time_zone_haiku::get_timezone()` on Haiku to ensure `Some` is returned and that the resulting string is non-empty.

### CODE REVIEW

File: haiku/src/implementation.cc [hunk 1]
```diff
@@ -57,6 +57,8 @@ size_t ::iana_time_zone_haiku::get_tz(rust::Slice<uint8_t> buf) {
 
 #else
 
-size_t ::iana_time_zone_haiku::get_tz(rust::Slice<uint8_t>) { return 0; }
```
Comment (lopopolo) [inline haiku/src/implementation.cc:60]: If I'm reading this set of docs correctly, this declaration syntax is only available since C++17: https://en.cppreference.com/w/cpp/language/namespace.

File: haiku/build.rs [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
-    println!("cargo:rustc-link-lib=be");
+    println!("cargo:return-if-changed=src/implementation.cc");
+    println!("cargo:return-if-changed=src/interface.h");
+
+    let target = env::var_os("TARGET").expect("cargo should set TARGET env var");
+    let target = target
+        .to_str()
+        .expect("TARGET env var should be valid UTF-8");
+    if target.contains("haiku") {
+        println!("cargo:rustc-link-lib=be");
```
Comment (lopopolo) [inline haiku/build.rs:18]: only link to `be` lib on haiku cross compile targets.

Because the build script runs on the host, we must parse the `TARGET` env variable that cargo sets.

File: haiku/src/implementation.cc (hunk 1 above)
Comment (Kijewski) [reply haiku/src/implementation.cc:60]: You're right! Could you do the same in line 13?

File: haiku/src/implementation.cc [hunk 3]
```diff
@@ -1,10 +1,10 @@
 #include "iana-time-zone-haiku/src/interface.h"
 #include "iana-time-zone-haiku/src/lib.rs.h"
 
-#ifdef __HAIKU__
-
 #include <cstring>
 
+#ifdef __HAIKU__
```
Comment (Kijewski) [inline haiku/src/implementation.cc:6]: Why did you move `#include <cstring>` outside of the guarded block?

File: haiku/src/implementation.cc (hunk 1 above)
Comment (lopopolo) [reply haiku/src/implementation.cc:60]: done

File: haiku/src/implementation.cc (hunk 3 above)
Comment (lopopolo) [reply haiku/src/implementation.cc:6]: because `size_t` is defined there which is needed for the `#else` block.

Comment (lopopolo) [reply haiku/src/implementation.cc]: gotcha. I undid these changes.

File: haiku/build.rs [hunk 4]
```diff
@@ -1,11 +1,20 @@
...
         .file("src/implementation.cc")
         .flag_if_supported("-std=c++11")
         .compile("tz_haiku");
 
-    println!("cargo:return-if-changed=include/interface.h");
-    println!("cargo:return-if-changed=src/implementation.cc");
     println!("cargo:return-if-changed=src/lib.rs");
-    println!("cargo:rustc-link-lib=be");
+    println!("cargo:return-if-changed=src/implementation.cc");
+    println!("cargo:return-if-changed=src/interface.h");
```
Comment (Kijewski) [inline haiku/build.rs:11]: Oops, thank you for catching that!

Comment (lopopolo) [reply haiku/build.rs]: thanks for commenting here. The build.rs directive is _rerun_ if changed, not return if changed. I pushed a new commit.

**Review by Kijewski — APPROVED** [review]

File: haiku/build.rs (hunk 4 above)
Comment (Kijewski) [reply haiku/build.rs]: Strange, I thought I copied the line instead of typing it out of memory. I have no idea where the typo came from.

**Comment by astraw** [comment]
@lopopolo this looks good.

Would it be possible to add configuration such that github actions does indeed run and build the haiku module, e.g. on macOS? Also, how does macOS actually factor in here? Is this the OS you developed this on?

**Comment by lopopolo** [comment]
@astraw I pushed some CI changes and pulled forward the warnings work from #76 specific to the haiku platform crate, fixed up docs, added another test.

**Comment by lopopolo** [comment]
I removed the `-Zminimal-versions` build for the haiku crate because `cxx-build` does not build with minimal versions. See:

- https://github.com/dtolnay/cxx/pull/1111

File: haiku/src/lib.rs [hunk 5]
```diff
@@ -27,3 +50,19 @@ pub fn get_timezone() -> Option<String> {
...
+    #[test]
+    #[cfg(not(target_os = "haiku"))]
+    fn test_fallback_on_non_haiku_platforms() {
+        assert!(super::get_timezone().is_none());
+    }
+
+    #[test]
+    #[cfg(target_os = "haiku")]
+    fn test_retrieve_time_zone_on_haiku_platforms() {
+        let timezone = super::get_timezone().unwrap();
```
Comment (lopopolo) [inline haiku/src/lib.rs:65]: I'm not sure if we should keep this test since the function is technically (and documented to be) fallible.

**Comment by lopopolo** [comment]
all the nightly android cross builds are failing. not sure why.

**Review by astraw — APPROVED** [review]

**Comment by lopopolo** [comment]
@astraw or @Kijewski can you publish this as `iana-time-zone-haiku` 0.1.1?

**Comment by Kijewski** [comment]
I'm not owner of iana-time-zone-haiku, only iana-time-zone. But I tested the current HEAD on the actual system (well, in a VM), and it works fine.

**Comment by astraw** [comment]
@Kijewski thanks. I just sent you an oversight to be publisher of iana-time-zone-haiku and also published it.
